scripts/periodic_cleanup.py

```diff
@@ -4,65 +4,152 @@
+        self.jenkins_client = jenkins_node.get_jenkins_client()
+        self.servers_by_region = dict()
+
+    @log
+    def read_env_vars(self):
+        self.age_limit = int(os.environ.get("INSTANCE_AGE_LIMIT", 48))
+        self.instance_prefix = os.environ.get("INSTANCE_PREFIX")
+        if not self.instance_prefix:
+            raise ValueError("INSTANCE_PREFIX must not be empty")
+
+        self.username = os.environ["PUBCLOUD_USERNAME"]
+        self.api_key = os.environ["PUBCLOUD_API_KEY"]
+        self.regions = os.environ["REGIONS"].split(' ')
+
+    @log
+    def init_rackspace_monitoring(self):
+        Cls = get_driver(Provider.RACKSPACE)
+        return Cls(self.username, self.api_key)
+
+    @log
+    def init_pyrax(self, region):
+        self.region = region
+        pyrax.set_setting("identity_type", "rackspace")
+        pyrax.set_credentials(self.username, self.api_key)
+        self.cs = pyrax.connect_to_cloudservers(
+            region, verify_ssl=True)
+
+    @log
+    def cache_servers(self):
+        self.servers = self.cs.servers.list()
+        self.servers_by_region[self.region] = self.servers
+        self.prefixed_servers = (
+            server for server in self.servers
+            if re.match(self.instance_prefix, server.name)
+        )
+
+    def get_servers_from_all_regions(self):
+        # flatten servers by region into list
+        return (s for sl in self.servers_by_region.values() for s in sl)
+
+    @log
+    def cleanup_instances(self):
+        """ Delete instances if they are in an error state or are over the
+        age defined in INSTANCE_AGE_LIMIT.
+
+        Instances that don't match INSTANCE_PREFIX are ignored.
+        """
+        current_time = datetime.datetime.now(tzutc())
+        max_age = datetime.timedelta(hours=self.age_limit)
+
+        for server in self.prefixed_servers:
+            created_time = dateutil.parser.parse(server.created)
+            age = current_time - created_time
+            errored = server.status == "ERROR"
+            if errored or age > max_age:
+                print("Deleting {name} Errored: {error} Age: {age}".format(
+                    name=server.name, error=errored, age=age))
+                server.delete()
+
+    @log
+    def cleanup_maas_entities(self):
+        """ Remove maas entities that relate to deleted instances """
+        entities = self.maas.list_entities()
+        server_names = (s.name for s in self.get_servers_from_all_regions())
+        for entity in entities:
+            # multi node aio entities are vm.hostname eg infra1.omnantp-52-66d1
+            # standard aio entities don't include a vm eg: ramcp-77-5dcd
+            hostname = entity.label.split('.')[-1]
+            # entities with URIs are tied to cloud servers and cannot be
+            # manually deleted
+            if (re.match(self.instance_prefix, hostname) and not entity.uri
+                    and hostname not in server_names):
```

Comment:
What happens if we deploy MaaS on a lab node?  The entities would get torn down here right?

Would it be better to check the agent `last_connected` value?  Not sure how this is exposed through the library, but from the cli:

```
<Agent: id=swift3.omnanxp-32-3cef, last_connected=1494909106476>
```

If we took that into account, we'd perhaps avoid deleting entities that shouldn't be.

One other small issue here ... we check a cached server list.  Completely unlikely, but something could come online after that list is generated, deploy entities, and then have them torn down because the host isn't in server_names.